src/array/cpu/spmm.h

```diff
@@ -0,0 +1,347 @@
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes. To avoid possible data hazard,
+ *       we use atomic operators in the reduction phase.
+ */
+template <typename IdType, typename DType, typename Op>
+void SpMMSumCoo(
+    const BcastOff& bcast,
+    const aten::COOMatrix& coo,
+    NDArray ufeat, NDArray efeat,
+    NDArray out) {
+  const bool has_idx = !aten::IsNullArray(coo.data);
+  const IdType* row = coo.row.Ptr<IdType>();
+  const IdType* col = coo.col.Ptr<IdType>();
+  const IdType* edges = coo.data.Ptr<IdType>();
+  const DType* X = ufeat.Ptr<DType>();
+  const DType* W = efeat.Ptr<DType>();
+  int64_t dim = bcast.out_len,
+          lhs_dim = bcast.lhs_len,
+          rhs_dim = bcast.rhs_len;
+  DType* O = out.Ptr<DType>();
+  const int64_t nnz = coo.row->shape[0];
+  // fill zero elements
+  memset(O, 0, out.GetSize());
+  // spmm
+#pragma omp parallel for
+  for (IdType i = 0; i < nnz; ++i) {
+    const IdType rid = row[i];
+    const IdType cid = col[i];
+    const IdType eid = has_idx? edges[i] : i;
+    DType* out_off = O + cid * dim;
+    for (int64_t k = 0; k < dim; ++k) {
+      const int64_t lhs_add = bcast.use_bcast ? bcast.lhs_offset[k] : k;
+      const int64_t rhs_add = bcast.use_bcast ? bcast.rhs_offset[k] : k;
+      const DType* lhs_off = Op::use_lhs? X + rid * lhs_dim + lhs_add : nullptr;
+      const DType* rhs_off = Op::use_rhs? W + eid * rhs_dim + rhs_add : nullptr;
+      const DType val = Op::Call(lhs_off, rhs_off);
+#pragma omp atomic
+      out_off[k] += val;
+    }
+  }
+}
+
+/*!
+ * \brief CPU kernel of SpMM-Min/Max on Csr format.
+ * \param bcast Broadcast information.
+ * \param csr The Csr matrix.
+ * \param ufeat The feature on source nodes.
+ * \param efeat The feature on edges.
+ * \param out The result feature on destination nodes.
+ * \param argu Arg-Min/Max on source nodes.
+ * \param arge Arg-Min/Max on edges.
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes.
+ */
+template <typename IdType, typename DType, typename Op, typename Cmp>
+void SpMMCmpCsr(
+    const BcastOff& bcast,
+    const aten::CSRMatrix& csr,
+    NDArray ufeat, NDArray efeat,
+    NDArray out, NDArray argu, NDArray arge) {
+  const bool has_idx = !aten::IsNullArray(csr.data);
+  const IdType* indptr = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices = static_cast<IdType*>(csr.indices->data);
+  const IdType* edges = has_idx ? static_cast<IdType*>(csr.data->data) : nullptr;
+  const DType* X = Op::use_lhs? static_cast<DType*>(ufeat->data) : nullptr;
+  const DType* W = Op::use_rhs? static_cast<DType*>(efeat->data) : nullptr;
+  const int64_t dim = bcast.out_len,
+                lhs_dim = bcast.lhs_len,
+                rhs_dim = bcast.rhs_len;
+  DType* O = static_cast<DType*>(out->data);
+  IdType* argX = Op::use_lhs? static_cast<IdType*>(argu->data) : nullptr;
```

Comment:
Is this for backward?
Can you add some doc about this?

Reply:
I have added some docstring describing arg_u and arg_e.